scene/main/viewport.cpp

```diff
@@ -2427,10 +2427,12 @@ void Viewport::_gui_control_grab_focus(Control *p_control) {
 		return;
 	}
 	get_tree()->call_group("_viewports", "_gui_remove_focus_for_window", (Node *)get_base_window());
-	gui.key_focus = p_control;
-	emit_signal(SNAME("gui_focus_changed"), p_control);
-	p_control->notification(Control::NOTIFICATION_FOCUS_ENTER);
-	p_control->queue_redraw();
+	if (p_control->is_inside_tree()) {
```

Comment:
As the MRP of #78590 shows, this crash can be caused by user-code. So the fix is necessary.
I suggest to make the condition even more robust by:
``` c++
if (p_control->is_inside_tree() && p_control->get_viewport() == this) {
```
Without this addition, user-code can bring the viewport into an inconsistent state, where a viewport has focus of a `Control`-node, that is not inside the viewport.

Here is a MRP, that demonstrates the necessary gdscript: [Test.zip](https://github.com/godotengine/godot/files/11851093/Test.zip)
In the MRP, follow the steps of #78590 and the label shows the path of the currently focused node in the SubViewport.

Reply:
You are right I didn't think about multiviewport I think only doing the second check should be enough as in case the node is out of the tree the viewport is a null object and should therefore never pass the second one.

Reply:
In a detached scene tree, `get_viewport()` still returns the (also detached) viewport.
I am not sure, if in this case the focus-change is desirable or not. I would tend to leave the `is_inside_tree` check in the code.

Reply:
I think this might cause the same type of bug where it can't find the viewport to unfocus himself if it is detached and reattached later. In fact if it does come early as we are treating it here but happening "normaly" it is detached so we should prevent from attaching if it happened before.

Reply:
I understand what you mean like if the viewport is removed from the scene tree it would unfocus the node so if the viewport has already been removed from scene tree `get_viewport` still return something but it should still prevent it to keep it consistent.